llvm/include/llvm/Target/GlobalISel/Combine.td

```diff
@@ -1642,7 +1652,7 @@ def all_combines : GICombineGroup<[trivial_combines, vector_ops_combines,
     sub_add_reg, select_to_minmax, redundant_binop_in_equality,
     fsub_to_fneg, commute_constant_to_rhs, match_ands, match_ors,
     combine_concat_vector, double_icmp_zero_and_or_combine, match_addos,
-    combine_shuffle_concat]>;
+    combine_shuffle_concat, combine_shuffle_to_extract_vector]>;
```

Comment:
What upstream checks are you referring to? I think the current `tryCombineShuffleVector` only matches merge-like patterns, but we could definitely move our `combine_shuffle_to_extract_vector` combine there, would that make it less expensive?

Reply:
No, checking elements to the shuffle vector is just an O(n²) operations unless you use a hashmap. At the moment, this is done in `tryCombineShuffleVector` which a backend _can_ enable. This is now run always in all backends. I don't know exactly why someone made the decision to have that combiner be run explicitly rather than using tablegen like the others